### ISSUE TITLE
Platform validation ensure/assert on RDG and RHI resource transitions and tracking [UE4.26.1]

### ISSUE THREAD

**Issue by jamiej** [NVIDIAGameWorks/RTXGI-DDGI#49]
### Question
Is there a forthcoming change better integrating with RDG making the manual RHI resource transitions observable to RDG validation?

### Issue
On one platform I'm seeing ensure/asserts identifying resources not transitioned appropriately. Unsure if this is platform specific validation bug. Nevertheless, I suspect none of the transition passes matter with respect to the validation error; there's no visibility between manual RHI resource transitions and RDG validation code as far as I can tell. 

Example log data from platform validation:

```
Line 20578: DDGIIrradiance: 0x12c0fd8110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x11610b9800
Line 20580: DDGIDistance: 0x12c102b110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x11610b9700
Line 20582: DDGIOffsets: 0x12c102ad10 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x11610b9500
Line 20584: DDGIStates: 0x12c1027110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x11610b9400
Line 20596: DDGIIrradiance: 0x12c0fd8110 (Whole Resource), Type: Assert, Usage: SRV_Compute, Trace: 0x11610b9200
Line 20611: DDGIDistance: 0x12c102b110 (Whole Resource), Type: Assert, Usage: SRV_Compute, Trace: 0x11610b9100
Line 20613: DDGIStates: 0x12c1027110 (Whole Resource), Type: Assert, Usage: SRV_Compute, Trace: 0x11610b8c00
Line 20615: DDGIOffsets: 0x12c102ad10 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x1161025300
Line 20617: DDGIIrradiance: 0x12c0fd8110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x11610be500
Line 20631: DDGIStates: 0x12c1027110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x1161024700
Line 20633: DDGIDistance: 0x12c102b110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x11625dd400
Line 20635: DDGIStates: 0x12c1027110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x1161023b00
Line 20638: DDGIIrradiance: 0x12c0fd8110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x116254b400
Line 20640: DDGIIrradiance: 0x12c0fd8110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x1161027000
Line 20642: DDGIDistance: 0x12c102b110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x1162544200
Line 20644: DDGIDistance: 0x12c102b110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x116103e200
Line 20646: DDGIOffsets: 0x12c102ad10 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x1162549b00
Line 20648: DDGIStates: 0x12c1027110 (Whole Resource), Type: Assert, Usage: UAV_Compute, Trace: 0x1161025c00
Line 20650: DDGIIrradiance: 0x12c0fd8110 (Whole Resource), Type: Assert, Usage: SRV_Graphics, Trace: 0x1162545d00
Line 20652: DDGIDistance: 0x12c102b110 (Whole Resource), Type: Assert, Usage: SRV_Graphics, Trace: 0x1161029200
Line 20654: DDGIStates: 0x12c1027110 (Whole Resource), Type: Assert, Usage: SRV_Graphics, Trace: 0x1162546b00
Line 20656: DDGIOffsets: 0x12c102ad10 (Whole Resource), Type: Assert, Usage: UAV_Graphics, Trace: 0x116100c600
Line 20658: DDGIDummyUAV: 0x12c142fd10 (Whole Resource), Type: Assert, Usage: UAV_Graphics, Trace: 0x116254e200
```

The above log represents load+update+diffuse render of a simple scene with 1 DDGIVolume. DDGIDummyUAV shows at tail representing the 5 remaining empty slots fill by the following code:

```
// When there are fewer relevant volumes than the maximum supported, set the empty volume texture slots to dummy values
for (int32 volumeIndex = numVolumes; volumeIndex < FDDGIVolumeSceneProxy::FComponentData::c_RTXGI_DDGI_MAX_SHADING_VOLUMES; ++volumeIndex)
{
	PassParameters->DDGIVolume[volumeIndex].ProbeIrradiance = GSystemTextures.BlackDummy->GetRenderTargetItem().TargetableTexture;
	PassParameters->DDGIVolume[volumeIndex].ProbeDistance = GSystemTextures.BlackDummy->GetRenderTargetItem().TargetableTexture;
	PassParameters->DDGIVolume[volumeIndex].ProbeOffsets = FDDGIVolumeSceneProxy::GetDummyUAV();
	PassParameters->DDGIVolume[volumeIndex].ProbeStates = GSystemTextures.BlackDummy->GetRenderTargetItem().TargetableTexture;
}
```

### Understanding of Code Flow
Resources are created, then Render Diffuse Lighting coordinates resource update w/ compositing scene result. Copy textures are performed once per volume, when asset data is loaded. Each update function operates on a subset of the volumes resources, the pass parameters identify which of those resources it touches, and finally may contain a transition pass inserting barriers for future passes. 

### Concluding Analysis
Primarily the validation asserts are due to combining both RDG and RHI directly managing barriers manually. This is caught by Platform logic checks, flagging the identified resources as not being accessible for specific pass units (compute/raster).

**Comment by jamiej** [comment]
Here's an emitted ensure from the platform:
```
Attempted to access resource "DDGIIrradiance" (0x12c106b910) (Whole Resource) from a hardware unit it is not currently accessible from. A resource transition is required.

    --- Allowed access states for this resource are: None.
    --- Compatible usages are: None.

    --- Required access states are: Write_ComputeUnit.
    --- Required usage is: UAV_Compute.
```

"None" can also be read as "Unknown", and that was the first clue RDG is unable to observe RHI transitions as performed in barrier passes.

**Comment by jamiej** [comment]
Verified it's not platform specific erroneous validation. 

Win64 platform using '-RHIValidation' command line option, which enables FVAlidationRHI
WindowsD3D12Device.cpp @ 492:
```
#if ENABLE_RHI_VALIDATION
	if (FParse::Param(FCommandLine::Get(), TEXT("RHIValidation")))
	{
		return new FValidationRHI(GD3D12RHI);
	}
#endif
```

Configuring auto logging for the resources of interest: 

Engine\Source\Runtime\RHI\Private\RHIValidation.cpp @ 25:
```
static const TCHAR* GAutoLogResourceNames[] =
{
	//
	// Add resource names here to automatically enable barrier logging for those resources.
	// e.g.
	//  TEXT("SceneDepthZ"),
	//
	TEXT("DDGIDistance"),
	TEXT("DDGIIrradiance"),
	TEXT("DDGIOffsets"),
	TEXT("DDGIStates"),
	TEXT("DDGIDummyUAV"),
	// ----------------------
	// Add names above this line
	nullptr
};
```

Results in the following log snippet:
```
WindowsNoEditor.log:1021:DDGIIrradiance: (0x000001FB78B60B60) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB792CE800
WindowsNoEditor.log:1027:DDGIDistance: (0x000001FB78B676A0) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB792CCA00
WindowsNoEditor.log:1033:DDGIOffsets: (0x000001FB78B613A0) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB79581B00
WindowsNoEditor.log:1039:DDGIStates: (0x000001FB78B6C660) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB792CE900
WindowsNoEditor.log:1134:DDGIIrradiance: (0x000001FB78B60B60) (Whole Resource), Type: Assert, Access: CopyDest, Pipeline: Graphics, Trace: 0x000001FB7958DA00
WindowsNoEditor.log:1136:DDGIDistance: (0x000001FB78B676A0) (Whole Resource), Type: Assert, Access: CopyDest, Pipeline: Graphics, Trace: 0x000001FB792CF000
WindowsNoEditor.log:1138:DDGIOffsets: (0x000001FB78B613A0) (Whole Resource), Type: Assert, Access: CopyDest, Pipeline: Graphics, Trace: 0x000001FB79580D00
WindowsNoEditor.log:1140:DDGIStates: (0x000001FB78B6C660) (Whole Resource), Type: Assert, Access: CopyDest, Pipeline: Graphics, Trace: 0x000001FB79580E00
WindowsNoEditor.log:1171:DDGIIrradiance: (0x000001FB78B60B60) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB0B6FD800
WindowsNoEditor.log:1173:DDGIStates: (0x000001FB78B6C660) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB0DFF5800
WindowsNoEditor.log:1175:DDGIDistance: (0x000001FB78B676A0) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB0DFF5900
WindowsNoEditor.log:1177:DDGIStates: (0x000001FB78B6C660) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB207B9700
WindowsNoEditor.log:1188:DDGIIrradiance: (0x000001FB78B60B60) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB0DFF5100
WindowsNoEditor.log:1190:DDGIIrradiance: (0x000001FB78B60B60) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB0DFF5000
WindowsNoEditor.log:1192:DDGIDistance: (0x000001FB78B676A0) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB0DFF4F00
WindowsNoEditor.log:1194:DDGIDistance: (0x000001FB78B676A0) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB0DFF4E00
WindowsNoEditor.log:1204:DDGIOffsets: (0x000001FB78B613A0) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB0DFF4900
WindowsNoEditor.log:1210:DDGIStates: (0x000001FB78B6C660) (Whole Resource), Type: Assert, Access: UAVCompute, Pipeline: Graphics, Trace: 0x000001FB0DFF4600
WindowsNoEditor.log:1216:DDGIIrradiance: (0x000001FB78B60B60) (Whole Resource), Type: Assert, Access: SRVGraphics, Pipeline: Graphics, Trace: 0x000001FB0DFF4300
WindowsNoEditor.log:1218:DDGIDistance: (0x000001FB78B676A0) (Whole Resource), Type: Assert, Access: SRVGraphics, Pipeline: Graphics, Trace: 0x000001FB7FCC9600
WindowsNoEditor.log:1220:DDGIStates: (0x000001FB78B6C660) (Whole Resource), Type: Assert, Access: SRVGraphics, Pipeline: Graphics, Trace: 0x000001FB7FCCA300
WindowsNoEditor.log:1222:DDGIOffsets: (0x000001FB78B613A0) (Whole Resource), Type: Assert, Access: UAVGraphics, Pipeline: Graphics, Trace: 0x000001FB7FCC6400
WindowsNoEditor.log:1224:DDGIDummyUAV: (0x000001FB47BB3CE0) (Whole Resource), Type: Assert, Access: UAVGraphics, Pipeline: Graphics, Trace: 0x000001FB7FCCA400
```

There are slight difference in what information is logged, but "Type: Assert" is consistent. With that done, will see if I can fix locally and share the result.

**Comment by acmarrs-nvidia** [comment]
First, thanks for this detailed analysis @jamiej. In the latest code, DDGIDummyUAV no longer exists so it won't be a problem anymore. Also, I made some updates for transitions to fix a few issues in Vulkan, but I am seeing the same RHI validation asserts that you've reported. I'll investigate. Thanks for pointing this out.

**Comment by acmarrs-nvidia** [comment]
These issues should be fixed in the latest 4.27 plugin release. Closing.
